server, ui: move the network latency report to the ui

BramGruneir opened the pull request:
Most of the work is in the ui.

<img width="813" alt="screen shot 2017-06-27 at 01 45 53" src="https://user-images.githubusercontent.com/1614265/27572529-8e4c3086-5ada-11e7-8c73-23baafa3afe2.png">

With a filter on node ids.
<img width="788" alt="screen shot 2017-06-27 at 01 46 08" src="https://user-images.githubusercontent.com/1614265/27572527-8e44c9c2-5ada-11e7-88a2-23872f92e6a1.png">

With two filters.
<img width="901" alt="screen shot 2017-06-27 at 01 46 31" src="https://user-images.githubusercontent.com/1614265/27572528-8e4954d8-5ada-11e7-985c-1e56e47cab63.png">

BramGruneir committed:
server, ui: move the network latency report to the ui

Most of the work is in the ui.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16728)
<!-- Reviewable:end -->

mrtracy commented:
Review status: 0 of 10 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/ui/src/views/reports/components/filterList/index.tsx, line 4 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnemCmAQWA3MNMtmYX_:-KnemCmAQWA3MNMtmYXa:blvvvsk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/components/filterList/index.tsx#L4)):*
> ```XML
> import _ from "lodash";
> 
> export interface FilterListProps {
> ```

Props interfaces do not need to be exported unless they are being used by another component somewhere.

---

*[pkg/ui/src/views/reports/components/filterList/index.tsx, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnelrLvdfnY0AcTxQ01:-KnelrLwP5HLN3afslkK:b-3mecw9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/components/filterList/index.tsx#L9)):*
> ```XML
> }
> 
> export function FilterList(props: FilterListProps) {
> ```

Consider using a name like "NodeFilterList" for this component; it's extremely specific to nodes, but the Reports module itself is not specific to nodes.

---

*[pkg/ui/src/views/reports/components/filterList/index.tsx, line 10 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnejTMKyqeYFQwNB-QJ:-KnejTMKyqeYFQwNB-QK:bw0ojgn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/components/filterList/index.tsx#L10)):*
> ```XML
> 
> export function FilterList(props: FilterListProps) {
>   const filters: string[] = [];
> ```

Using a destructing call at the beginning of the function often makes the function more readable, by cutting down on the number of times `props` needs to be referenced.

```typescript
const { nodeIDs, localityRegex } = props;
```

---

*[pkg/ui/src/views/reports/components/filterList/index.tsx, line 11 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KneiFkBpTj2U_5uyDll:-KneiFkCTL1BUGdyXDFw:b-fgminf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/components/filterList/index.tsx#L11)):*
> ```XML
> export function FilterList(props: FilterListProps) {
>   const filters: string[] = [];
>   if (!_.isNil(props.nodeIDs) && props.nodeIDs.size > 0) {
> ```

`!_.isEmpty(props.nodeIDs)`

---

*[pkg/ui/src/views/reports/components/filterList/index.tsx, line 12 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KneitfK5QjjVYtz6FKB:-KneitfK5QjjVYtz6FKC:bb9mamj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/components/filterList/index.tsx#L12)):*
> ```XML
>   const filters: string[] = [];
>   if (!_.isNil(props.nodeIDs) && props.nodeIDs.size > 0) {
>     const nodeList = _.join(_.map(Array.from(props.nodeIDs.keys()).sort(), (nodeID) => `n${nodeID}`), ", ");
> ```

This would be more readable if expressed with a `chain` wrapper.

```typescript
const nodeList = _.chain(props.nodes)
    .keys()
    .sort()
    .map((nid) => `${nid}`)
    .join(",");
```

The final "join" works directly on the chain wrapper, so you don't have to call `.value()` to unwrap it.

---

*[pkg/ui/src/views/reports/components/filterList/index.tsx, line 20 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnemK_RTTwFLKrvpBJR:-KnemK_RTTwFLKrvpBJS:b-bjs287) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/components/filterList/index.tsx#L20)):*
> ```XML
>   if (_.isEmpty(filters)) {
>     return null;
>   }
> ```

Add a newline here; some spaces to break up the code are often helpful in reading.

---

*[pkg/ui/src/views/reports/components/filterList/index.tsx, line 21 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnemU_xwatj9-vDHP5v:-KnemU_xwatj9-vDHP5w:b112x1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/components/filterList/index.tsx#L21)):*
> ```XML
>     return null;
>   }
>   return <div>
> ```

This starts a multiline JSX statement, put it in parenthesis.

```typescript
return (
  <div>
    ...
  </div>
);
```

---

*[pkg/ui/src/views/reports/components/filterList/index.tsx, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnelWkhDigVn2_feFqn:-KnelWkhDigVn2_feFqo:b-vxni3k) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/components/filterList/index.tsx#L23)):*
> ```XML
>   return <div>
>     <h2>Filters</h2>
>     <ul className="filter-ul">
> ```

Don't name the class after the specific element used.  This should also probably be more specific, "filter" is likely to make semantic sense for a completely different element on some other page.  Use something like "node-filter-list"

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnesVUuZzD24_J8ZuCA:-KnesVUuZzD24_J8ZuCB:b854vk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L25)):*
> ```XML
> 
> function getNodeIDs(input: string) {
>   const ids: Map<number, {}> = new Map();
> ```

If you're going to use Map(), you should just use Set() for this purpose.  They're both ES6 objects.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 37 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnemjOAby9lSb3ekVIn:-KnemjOBRDyAGbR2NgG7:bieljno) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L37)):*
> ```XML
> }
> 
> function nsToMs(ns: Long) {
> ```

Use "NanoToMilli" from the convert library.

```typescript
import { NanoToMilli } form "src/util/convert";
```

It works on `number`, not `Long`, and thus should be called as `NanoToMilli(longVal.toNumber()`.  The conversion to `number` does lose a small amount of precision for very large Longs, but we've already committed to a larger precision loss by converting to milliseconds in the first place.

If you're not satisfied with that, you could convert NanoToMilli to accept `Long` as well as `number`, which is easily done in typescript/javascript.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-Knevx77nx8eE02diKJS:-Knevx77nx8eE02diKJT:b-p0y4gw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L52)):*
> ```XML
>     props.refreshNodes(new protos.cockroach.server.serverpb.NodesRequest({
>       node_ids: (!_.isEmpty(props.location.query.node_ids)) ? props.location.query.node_ids : "",
>       locality: (!_.isEmpty(props.location.query.locality)) ? props.location.query.locality : "",
> ```

This is old code, should be removed; we're always querying all of the statuses.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 87 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KneuPh_xHc_ttnNluiP:-KneuPhaBEXLjkO7TetL:b-jauvry) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L87)):*
> ```XML
>       }
>     }
>     const lastUpdatedAt = _.maxBy(nodes, _ => "updated_at").updated_at;
> ```

This doesn't provide type safety, use `(n) => n.updated_at`.

Also, you shouldn't use this, what if every node in your subquery is down?  The actual nodes cachedReducer has a "last updated" value which is the last time it actually queried for data.  See `nodes.ts` for an example of doing this with liveness values.

Actually, you should probably just be using the liveness values from the nodeSummary provided by nodes.ts; they seem to use similar cutoffs.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 145 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-Knf7cAXxECom-RNEMeS:-Knf7cAXxECom-RNEMeT:b123vs5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L145)):*
> ```XML
>     let latencySum = 0;
>     let latencySquareSum = 0;
>     _.forEach(ids, (id) => {
> ```

Most of these calculations should be memoized selectors, not part of the render method.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728)*
<!-- Sent from Reviewable.io -->

couchand commented:
Reviewed 3 of 10 files at r1.
Review status: 3 of 10 files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 8 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnfKst-EeN3DwaUbA5l:-KnfKst-EeN3DwaUbA5m:b-eoj9pt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L8)):*
> ```XML
> import { AdminUIState } from "src/redux/state";
> import { refreshNodes } from "src/redux/apiReducers";
> import { RouterState } from "react-router";
> ```

Is there guidance on import ordering?  This might look nicer if modules imports were uniformly up above.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 83 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnfV6aud_A7rb6Vdjcv:-KnfV6aud_A7rb6Vdjcw:b2nicf1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L83)):*
> ```XML
>         locality = new RegExp(this.props.location.query.locality);
>       } catch (e) {
>         // ignore there error
> ```

sp?

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnfYMp_b-O7THcJ8VOO:-KnfYMp_b-O7THcJ8VOP:bh08usp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L101)):*
> ```XML
>     let ids: Identity[] = [];
>     // Map of latencies by starting node id.
>     const latenciesByNodeID: Map<number, { [k: string]: Long }> = new Map();
> ```

It seems a little strange that these wouldn't be the same data structure at both levels, since it seems the key space is identical.  My preference would be generally be for plain objects at both levels, but it seems whether they're Maps or objects-as-map they should be the same.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 145 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-Knf7cAXxECom-RNEMeS:-KnfZImtGchNlC8OAQCZ:bdczl12) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L145)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Most of these calculations should be memoized selectors, not part of the render method.
</blockquote></details>

And maybe just use the well-tested methods from D3? https://github.com/d3/d3-array/blob/master/README.md#statistics

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 234 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnfZiXvDhktvKzkJZ0f:-KnfZiXwRAhWVCcVbaDD:b-e61uvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L234)):*
> ```XML
>       }
>       const className = `network-table__cell network-table__cell--${heat}`;
>       const title = `n${nodeIDa} -> n${nodeIDb}\n${latency.toString()}ms`;
> ```

Should the latency here also be `.toFixed(2)`?

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 241 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnfZvhNXZB-TOU3LMaY:-KnfZvhNXZB-TOU3LMaZ:bi1pfst) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L241)):*
> ```XML
> 
>     // createHeaderCell creates and decorates a header cell.
>     function createHeaderCell(id: Identity, key: number, left: boolean = false) {
> ```

I have a personal vendetta against magic boolean parameters, it's almost never obvious from the call site what it means.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 299 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-Knf_L-tC3_4p5aRQGXg:-Knf_L-tC3_4p5aRQGXh:be71qka) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L299)):*
> ```XML
> 
>     // latencyTable is the table and heat-map that's displayed for all nodes.
>     const latencyTable = (_.isEmpty(ids)) ?
> ```

I also have a personal vendetta against multi-line ternaries, they're usually pretty hard to read & change safely.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 307 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-Knf_VYujVX__OqSETIo:-Knf_VYvv8wHMUsE5ZwJ:b-hu9fxe) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L307)):*
> ```XML
>         <table className="network-table">
>           <tbody>
>             <tr key="0" className="network-table__row">
> ```

I don't think you need keys for these static elements, please correct me if I'm wrong?

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 340 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-Knf_vAfhqGnJTZ8PdV2:-Knf_vAfhqGnJTZ8PdV3:b-hri4yz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L340)):*
> ```XML
>               </td>
>               <td className="network-table__cell network-table__cell--header">
>                 &lt; -1 stddev
> ```

Technically this label is wrong, it's actually between 1 and 2 stddev.  Not sure if that's worth trying to correct, since it's probably obvious from the context.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 441 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnfaUtNMC9Ida3oubzA:-KnfaUtNMC9Ida3oubzB:blssj3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L441)):*
> ```XML
> }
> 
> export default connect(
> ```

I generally like to define the `mapStateToProps` function above so the connect call is one line.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 442 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnfaLz4JSDuwRp9ZUXG:-KnfaLz4JSDuwRp9ZUXH:bv966jx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L442)):*
> ```XML
> 
> export default connect(
>   (state: AdminUIState) => {
> ```

This could be written in shorthand with parens like `(state) => ({ nodes: data })`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 1 of 10 files reviewed at latest revision, 23 unresolved discussions, some commit checks pending.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 8 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnfKst-EeN3DwaUbA5l:-KnkDAlqNap4kMjLK7vC:b-cgvmnu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L8)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Is there guidance on import ordering?  This might look nicer if modules imports were uniformly up above.
</blockquote></details>

@mrtracy?  What is the guidance here.  I agree that we should have some strict ordering.  I've ordered these as external then internal, ordered alphabetically by final source file and then by import name.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnesVUuZzD24_J8ZuCA:-KnkDLBw7198waUxuUNt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L25)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

If you're going to use Map(), you should just use Set() for this purpose.  They're both ES6 objects.
</blockquote></details>

Done.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 37 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnemjOAby9lSb3ekVIn:-KnkDkhxjN21svTeZWIJ:bv3mq49) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L37)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Use "NanoToMilli" from the convert library.

```typescript
import { NanoToMilli } form "src/util/convert";
```

It works on `number`, not `Long`, and thus should be called as `NanoToMilli(longVal.toNumber()`.  The conversion to `number` does lose a small amount of precision for very large Longs, but we've already committed to a larger precision loss by converting to milliseconds in the first place.

If you're not satisfied with that, you could convert NanoToMilli to accept `Long` as well as `number`, which is easily done in typescript/javascript.
</blockquote></details>

Done.  

It would be nice to add the long version sometime.  But not now.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-Knevx77nx8eE02diKJS:-KnkDuhPpF_1NHNe7oFR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L52)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

This is old code, should be removed; we're always querying all of the statuses.
</blockquote></details>

Done.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 83 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnfV6aud_A7rb6Vdjcv:-KnkE2momOX6k1qC_0z-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L83)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

sp?
</blockquote></details>

Done.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 87 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KneuPh_xHc_ttnNluiP:-KnkE4b50IA0v0oxaePc:b-sxrv66) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L87)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

This doesn't provide type safety, use `(n) => n.updated_at`.

Also, you shouldn't use this, what if every node in your subquery is down?  The actual nodes cachedReducer has a "last updated" value which is the last time it actually queried for data.  See `nodes.ts` for an example of doing this with liveness values.

Actually, you should probably just be using the liveness values from the nodeSummary provided by nodes.ts; they seem to use similar cutoffs.
</blockquote></details>

Converted to the liveness values... BUT ordering by multiple values is a little bit odd.  Here's the most egregious case.

```
       ... }))
        .sortBy(noConnection => noConnection.to.nodeID)
        .sortBy(noConnection => noConnection.to.locality)
        .sortBy(noConnection => noConnection.from.nodeID)
        .sortBy(noConnection => noConnection.from.locality)
```

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnfYMp_b-O7THcJ8VOO:-KnkEofjkeSyYHi26dOa:b5psy4y) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L101)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

It seems a little strange that these wouldn't be the same data structure at both levels, since it seems the key space is identical.  My preference would be generally be for plain objects at both levels, but it seems whether they're Maps or objects-as-map they should be the same.
</blockquote></details>

no need for that map anymore, but the idea was to not convert all the underlying data where there was no need.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 145 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-Knf7cAXxECom-RNEMeS:-KpGiesipFI343gG4A3X:b-b7843b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L145)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

And maybe just use the well-tested methods from D3? https://github.com/d3/d3-array/blob/master/README.md#statistics
</blockquote></details>

Changed this to use the D3 methods.  I'll create a selector in a follow up PR.  Added a todo as this PR would seem a little bloated.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 234 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnfZiXvDhktvKzkJZ0f:-KnkF0tgmMlp1pjr3-BW:bdw4m30) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L234)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Should the latency here also be `.toFixed(2)`?
</blockquote></details>

This is intentional. so that the title has the full number instead of the limited one used in the heatmap.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 241 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnfZvhNXZB-TOU3LMaY:-KpGvc3C9CCyETe7XO_0:b-d9q7x1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L241)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I have a personal vendetta against magic boolean parameters, it's almost never obvious from the call site what it means.
</blockquote></details>

Done.  Moved this into CSS where it should be.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 299 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-Knf_L-tC3_4p5aRQGXg:-KpGQX1yJ1HH_e3aIIrj:br7lky6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L299)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I also have a personal vendetta against multi-line ternaries, they're usually pretty hard to read & change safely.
</blockquote></details>

I agree.  Done.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 307 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-Knf_VYujVX__OqSETIo:-KpGQ96WbPe7vLKtDNfj:bcfwjge) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L307)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I don't think you need keys for these static elements, please correct me if I'm wrong?
</blockquote></details>

 Is it needed?  Maybe not, but it ensures that all the rows have a key, since the map afterwards has keys for all the following rows.  Seemed strange not to have a key for the first row but one for all the rest.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 340 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-Knf_vAfhqGnJTZ8PdV2:-KpGQ_n5mEZtOXphNLVM:b-yrf00u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L340)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Technically this label is wrong, it's actually between 1 and 2 stddev.  Not sure if that's worth trying to correct, since it's probably obvious from the context.
</blockquote></details>

Yeah, it's quite clear.  I'm going to leave this as is unless there are move complaints.  I had the longer label in there during some earlier iteration, but it just made it less readable.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 441 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnfaUtNMC9Ida3oubzA:-KpGvinDr_nFn01n0d_m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L441)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I generally like to define the `mapStateToProps` function above so the connect call is one line.
</blockquote></details>

Done.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 442 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnfaLz4JSDuwRp9ZUXG:-KpGvj9aI27rN87CaSg4:b-fqu785) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L442)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

This could be written in shorthand with parens like `(state) => ({ nodes: data })`
</blockquote></details>

Can't do that in this case.  

---

*[pkg/ui/src/views/reports/components/filterList/index.tsx, line 4 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnemCmAQWA3MNMtmYX_:-KnkBGBDazCcV6O-mOk-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/components/filterList/index.tsx#L4)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Props interfaces do not need to be exported unless they are being used by another component somewhere.
</blockquote></details>

Done.

---

*[pkg/ui/src/views/reports/components/filterList/index.tsx, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnelrLvdfnY0AcTxQ01:-KnkBOqt8VHriAyq8fcx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/components/filterList/index.tsx#L9)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Consider using a name like "NodeFilterList" for this component; it's extremely specific to nodes, but the Reports module itself is not specific to nodes.
</blockquote></details>

Done.

---

*[pkg/ui/src/views/reports/components/filterList/index.tsx, line 10 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnejTMKyqeYFQwNB-QJ:-KnkC3IG3eCK0tifN5TW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/components/filterList/index.tsx#L10)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Using a destructing call at the beginning of the function often makes the function more readable, by cutting down on the number of times `props` needs to be referenced.

```typescript
const { nodeIDs, localityRegex } = props;
```
</blockquote></details>

Done.

---

*[pkg/ui/src/views/reports/components/filterList/index.tsx, line 11 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KneiFkBpTj2U_5uyDll:-KnkC9FBdYhtkAriMh3G:b-48qqyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/components/filterList/index.tsx#L11)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

`!_.isEmpty(props.nodeIDs)`
</blockquote></details>

Sadly, doesn't work with maps or sets.  None of lodash does. 

---

*[pkg/ui/src/views/reports/components/filterList/index.tsx, line 12 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KneitfK5QjjVYtz6FKB:-KnkCdv7HE1E6d0zmzLQ:by1erov) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/components/filterList/index.tsx#L12)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

This would be more readable if expressed with a `chain` wrapper.

```typescript
const nodeList = _.chain(props.nodes)
    .keys()
    .sort()
    .map((nid) => `${nid}`)
    .join(",");
```

The final "join" works directly on the chain wrapper, so you don't have to call `.value()` to unwrap it.
</blockquote></details>

Done.  I love these chains.

---

*[pkg/ui/src/views/reports/components/filterList/index.tsx, line 20 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnemK_RTTwFLKrvpBJR:-KnkD-pQaORIiGLUGoLX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/components/filterList/index.tsx#L20)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Add a newline here; some spaces to break up the code are often helpful in reading.
</blockquote></details>

Done.

---

*[pkg/ui/src/views/reports/components/filterList/index.tsx, line 21 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnemU_xwatj9-vDHP5v:-KnkD1GvKa16D6cOtKCI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/components/filterList/index.tsx#L21)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

This starts a multiline JSX statement, put it in parenthesis.

```typescript
return (
  <div>
    ...
  </div>
);
```
</blockquote></details>

Done.

---

*[pkg/ui/src/views/reports/components/filterList/index.tsx, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnelWkhDigVn2_feFqn:-KnkD9DD3_oIiwOCiivu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/components/filterList/index.tsx#L23)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Don't name the class after the specific element used.  This should also probably be more specific, "filter" is likely to make semantic sense for a completely different element on some other page.  Use something like "node-filter-list"
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Lots of changes to this, PTAL.

---

Review status: 1 of 10 files reviewed at latest revision, 23 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-:-KpGwBz1Ma0y_J6xx5p3:b-l3vxuz)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
This is in a better place and I'm LGTM; my one concern is that, by the end of the `render` method of the main page component, there are a large number of variables in scope and it's a bit confusing to follow.  But honestly, i'm not sure it's enough variables to break it up into components.

---

Review status: 1 of 10 files reviewed at latest revision, 24 unresolved discussions, all commit checks successful.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 8 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KnfKst-EeN3DwaUbA5l:-KpIVH9ySWfcmRO1_bTu:byno4ig) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L8)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

@mrtracy?  What is the guidance here.  I agree that we should have some strict ordering.  I've ordered these as external then internal, ordered alphabetically by final source file and then by import name.
</blockquote></details>

I've been doing external/internal/relative, haven't been enforcing anything though.  There is a linter rule to alphabetize within groups, but no linter rules available to enforce the groups themselves.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 87 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KneuPh_xHc_ttnNluiP:-KpIiK5O1JxEWF2WZfkQ:b-3pvccr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L87)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Converted to the liveness values... BUT ordering by multiple values is a little bit odd.  Here's the most egregious case.

```
       ... }))
        .sortBy(noConnection => noConnection.to.nodeID)
        .sortBy(noConnection => noConnection.to.locality)
        .sortBy(noConnection => noConnection.from.nodeID)
        .sortBy(noConnection => noConnection.from.locality)
```
</blockquote></details>

sortBy should take a list of criteria.
```
.sortBy([
    noConnection => noConnection.to.nodeID,
    noConnection => noConnection.to.locality,
    noConnection => noConnection.from.nodeID,
    noConnection => noConnection.from.locality,
])
```

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 307 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-Knf_VYujVX__OqSETIo:-KpIWwlgo4tQtkicjyCR:b-sibv48) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L307)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

 Is it needed?  Maybe not, but it ensures that all the rows have a key, since the map afterwards has keys for all the following rows.  Seemed strange not to have a key for the first row but one for all the rest.
</blockquote></details>

The key is a React requirement when rendering a list of elements.  React can use it to do efficient DOM re-renders when the list is modified in certain ways. These optimizations are not needed for static elements, so they don't need keys (and shouldn't have them). 

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 19 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KpIazs7Op5lnXomkZtp:-KpIazs7Op5lnXomkZtq:b-4yr6dt) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4658d93376bc5b766cb7cd9491f151e27e2b25c/pkg/ui/src/views/reports/containers/network/index.tsx#L19)):*
> ```XML
>   refreshNodes: typeof refreshNodes;
>   refreshLiveness: typeof refreshLiveness;
> }
> ```

newline

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-:-KpIxSoKIk8A5mdfUAmC:bzgmouj)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 1 of 10 files reviewed at latest revision, 24 unresolved discussions.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 87 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KneuPh_xHc_ttnNluiP:-KpL7jSvqfYaY2VZO8g9:br6rrrn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L87)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

sortBy should take a list of criteria.
```
.sortBy([
    noConnection => noConnection.to.nodeID,
    noConnection => noConnection.to.locality,
    noConnection => noConnection.from.nodeID,
    noConnection => noConnection.from.locality,
])
```
</blockquote></details>

Nope :(

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 307 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-Knf_VYujVX__OqSETIo:-KpL8Db7dBqgXsTnGuTx:btfxb5v) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L307)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

The key is a React requirement when rendering a list of elements.  React can use it to do efficient DOM re-renders when the list is modified in certain ways. These optimizations are not needed for static elements, so they don't need keys (and shouldn't have them). 
</blockquote></details>

Ok, removed.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 19 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KpIazs7Op5lnXomkZtp:-KpL7NqbuD3ILiNPK90k:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4658d93376bc5b766cb7cd9491f151e27e2b25c/pkg/ui/src/views/reports/containers/network/index.tsx#L19)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

newline
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 1 of 10 files reviewed at latest revision, 24 unresolved discussions, some commit checks pending.

---

*[pkg/ui/src/views/reports/containers/network/index.tsx, line 87 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728#-KneuPh_xHc_ttnNluiP:-KpLLUCuTLw2zPUfOvRs:b-s9pcfi) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cd2d9c24f77cf776a5307e89c18bfd6c239c0b/pkg/ui/src/views/reports/containers/network/index.tsx#L87)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Nope :(
</blockquote></details>

Apparently `sortBy` doesn't have that signature when used as part of a chain wrapper. Oh well, the way you have it is fine for now, it's readable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16728)*
<!-- Sent from Reviewable.io -->
